### ISSUE TITLE
Passing QueryParameters during rerouting in a BeforeEnterEvent

### ISSUE THREAD

**Issue by tarekoraby** [vaadin/flow#8864]
In V14.3.3, the docs don't clarify the way by which one can dynamically reroute to a different UI using `BeforeEnterEvent` while also passing along any query parameters that the current route has. For example, it's not clear how can one redirect `firstView?foo=bar` to `secondView?foo=bar` using a `BeforeEnterEvent` (if that's at all possible).

To illustrate further, consider the following two views.

```
@Route("firstView")
public class FirstView extends VerticalLayout
        implements BeforeEnterObserver {


    public FirstView() {
        add(new H3("First view"));
    }

    @Override
    public void beforeEnter(BeforeEnterEvent event) {
        event.rerouteTo("secondView");
        // event.rerouteTo("secondView", "someParameter");
    }
}
```

And ... 

```
@Route("secondView")
public class SecondView extends VerticalLayout implements HasUrlParameter<String>  {

    Span span;

    public SecondView() {
        add(new H3("Second view"));
        span = new Span("none");
        add(new H3("QueryParameters:"), span);
    }

    @Override
    public void setParameter(BeforeEvent event, @OptionalParameter String parameter) {

        Location location = event.getLocation();
        QueryParameters queryParameters = location
                .getQueryParameters();

        Map<String, List<String>> parametersMap =
                queryParameters.getParameters();

        span.setText(parametersMap.toString());

    }
}
```

If based on these views, one would browse to `firstView?foo=bar` then the rerouting to `secondView` would occur without passing the `QueryParameters`. 

Strangely, if instead of rerouting using `BeforeEvent.rerouteTo(String route)`, one uses `BeforeEvent.rerouteTo(String route, T routeParam)` then the `QueryParameters `are indeed passed to the other view (even though the passed 'routeParam' has nothing to do AFAIK with the passed `QueryParameters`).

**Comment by caalador** [comment]
The Query parameters are actually always sent to the forwarded and redirected views.
In this case the setParameter method for `OptionalParameter` doesn't fire for a null parameter at all, I would say that is a bug and the reason this is not working as expected. 
If you change the `HasUrlParameter` to `BeforeEnterObserver` you'll note that the query parameters are there.

**Comment by tarekoraby** [comment]
Thanks for the clarification!

> Query parameters are actually always sent to the forwarded and redirected views

Is it worth clarifying this point in the docs? And should a bug label be added for the setParameter() not being fired?

**Comment by caalador** [comment]
The original issue was #7188 and fixed in #7189
A quick test notes that the setParameter is fired for no parameter when going directly to the page, but will not 
be set when coming form a redirect if no parameter is set. This should get it's own ticket and this closed.
Could you collect the information for the new ticket and close this one?
Else I'll try to make it tomorrow.

**Comment by caalador** [comment]
Created issue #8867 for the actual cause. Thanks for this issue that did lead to a bug.
I'm closing this issue, but if you feel that queryparameter retention needs a specific mention in documentation
please open a ticket to https://github.com/vaadin/flow-and-components-documentation/.
